### PR TITLE
List sources and Get source using Sources Client

### PR DESCRIPTION
**Pull request by adiabramovitch**
Depends on https://github.com/RHEnVision/provisioning-backend/pull/31

### CODE REVIEW

File: internal/routes/all_routes.go [hunk 1]
```diff
@@ -39,5 +39,12 @@ func apiRouter() http.Handler {
 		})
 	})
 
+	r.Route("/sources", func(r chi.Router) {
+		r.Get("/", s.ListSources)
+		r.Route("/{ID}", func(r chi.Router) {
+			r.Get("/", s.GetSource)
+		})
+	})
+
```
Comment (adiabramovitch) [inline internal/routes/all_routes.go:48]: Need to decide if sources/aws or sources?provider=aws

**Comment by lzap** [comment]
I've created a PR which slightly modifies the Makefile so tell me what you think: https://github.com/RHEnVision/provisioning-backend/pull/39

I see a pattern for configuration: for each REST API we will need a configurable option, so perhaps it is worth creating a common configuration group:

```yaml
restEndpoints:
  imageBuilder:
    url: xxxx
  sources:
    url: xxxx
    otherOption: xxxx
```

We can only export RestEndpoints struct, no need to export all values.

**Comment by adiabramovitch** [comment]
> 

Sounds good to me, we can always change it if ou would need more information about the client.

File: internal/routes/all_routes.go (hunk 1 above)
Comment (adiabramovitch) [reply internal/routes/all_routes.go]: yes I agree

**Comment by lzap** [comment]
There are conflicts FYI.

File: internal/payloads/sources_payload.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import (
+	"net/http"
+
+	"github.com/RHEnVision/provisioning-backend/internal/clients/sources"
+	"github.com/go-chi/render"
+)
+
+type ShowSourceResponse struct {
+	sources.Source
+}
```
[the remark this answers is not in the document]
Comment (lzap) [reply internal/payloads/sources_payload.go:12]: Also I would use pointer in here, generally speaking not using pointers is only good for small structs as a very tight optimalization technique.

File: internal/db/seeds/dev_small.sql [hunk 3]
```diff
@@ -5,7 +5,7 @@
 BEGIN;
 
 INSERT INTO accounts(id, account_number, org_id)
-VALUES (1, '13', '000013'),
+VALUES (1, '6089719', '000013'),
```
Comment (adiabramovitch) [inline internal/db/seeds/dev_small.sql:8]: I have changed it in order to authenticate both sources and our service :)

**Comment by ezr-ondrej** [comment]
Are we adding tests, or shall that be a follow-up (which I'm totally fine with)? :)

File: internal/clients/sources/sources_client.gen.go [hunk 4]
```diff
@@ -1,7 +1,7 @@
 // Package SourcesApi provides primitives to interact with the openapi HTTP API.
 //
 // Code generated by unknown module path version unknown version DO NOT EDIT.
-package SourcesApi
+package sources
 
 import (
 	"bytes"
```
Comment (lzap) [inline internal/clients/sources/sources_client.gen.go:7]: Please do not make manual edits to generate files. Remove it from this patch, I will do a PR to fix this. It looks like generation does not work correctly even.

**Review by lzap — CHANGES_REQUESTED** [review]
Lookin' good!

File: internal/services/sources_service.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	resp, err := client.ListSourcesWithResponse(r.Context(), &sources.ListSourcesParams{}, AddIdentityHeader)
+	if err != nil {
+		// TODO: use the errors supported by the JSON 400 OR 404
+		renderError(w, r, payloads.NewInvalidRequestError(r.Context(), err))
+		return
+	}
+
+	if err := render.RenderList(w, r, payloads.NewListSourcesResponse(resp.JSON200.Data)); err != nil {
+		// TODO: use the errors supported by the JSON 400 OR 404
+		renderError(w, r, payloads.NewRenderError(r.Context(), "list sources", err))
```
[the remark this answers is not in the document]
Comment (adiabramovitch) [reply internal/services/sources_service.go:36]: No, In the resp struct there are also errors documented which I want to show as well on a different PR.

File: internal/services/sources_service.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+		return
+	}
+
+}
+
+func GetSource(w http.ResponseWriter, r *http.Request) {
+	id := chi.URLParam(r, "ID")
```
[the remark this answers is not in the document]
Comment (adiabramovitch) [reply internal/services/sources_service.go:43]: I want it as a string not an int

File: internal/clients/sources/sources_client.gen.go (hunk 4 above)
Comment (lzap) [reply internal/clients/sources/sources_client.gen.go:7]: FYI the PR is here, WIP: https://github.com/RHEnVision/provisioning-backend/pull/46

File: internal/services/sources_service.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (lzap) [reply internal/services/sources_service.go:43]: Sorry disregard.

**Comment by adiabramovitch** [comment]
@lzap , @ezr-ondrej I will open today PRs for adding more error handling and the UI dropdown. 
Regarding tests, I'll add those as well in a different PR I want the PRs to be as small as possible. 

@lzap  I addressed your comments.

**Review by ezr-ondrej — COMMENTED** [review]
I do not know how to test a happy path (how do I seed sources data?), but when sources return 404, we panic...

File: internal/db/seeds/dev_small.sql [hunk 7]
```diff
@@ -8,7 +8,8 @@ INSERT INTO accounts(id, account_number, org_id)
 VALUES (1, '13', '000013'),
        (2, '15', '000015'),
        (3, NULL, '000042'),
-       (4, NULL, '000077')
+       (4, NULL, '000077'),
+       (5, '6089719', '000016')
```
Comment (adiabramovitch) [inline internal/db/seeds/dev_small.sql:12]: You need to make sure you have this org_id linked to this account number in sources. 
(It is there on default) @ezr-ondrej

Comment (lzap) [reply internal/db/seeds/dev_small.sql:12]: Sure makes sense, but not all our endpoints will interact with sources. Our own accounts still make sense.

File: internal/services/sources_service.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+func ListSources(w http.ResponseWriter, r *http.Request) {
+	client, err := sources.NewClientWithResponses(*config.SourcesURL)
+	if err != nil {
+		renderError(w, r, payloads.NewClientInitializationError(r.Context(), "sources client", err))
+		return
+	}
+	resp, err := client.ListSourcesWithResponse(r.Context(), &sources.ListSourcesParams{}, AddIdentityHeader)
+	if err != nil {
+		// TODO: use the errors supported by the JSON 400 OR 404
```
Comment (lzap) [inline internal/services/sources_service.go:21]: What does this mean? Obviously we want to use 404 when a resource is not found, which is probably not the case for listing operation.

The code you have is good, you return 3rd party error when unknown error occurs which is HTTP 500 which is correct I think. That is the code for "unknown backend errors".

**Review by lzap — DISMISSED** [review]
It's perfect, one nitpick about the "New" in the error message.

**Review by ezr-ondrej — APPROVED** [review]
Let's get it in, lets get the filtering by provider and tests in follow ups :+1:

**Pull request merged**